### DISCUSSION TITLE
Debouncing event sent to invoked callback service

### DISCUSSION THREAD

**Discussion by matigda** [statelyai/xstate#1596] [Q&A]
Hi, I've been working for some time with xState and it's awesome tool! However right now I'm struggling with case I can't really find any solutions for. It's canceling events on invoked callback service. I think code will explain better my intention. Below I just copied the important parts, to make it easier for you to understand.

```
Machine
({
    id: 'checkout',
     ....
    type: 'parallel',
    states: {
        items: { 
            initial: "idle",
            invoke: {
                 src: 'itemsService'
             },
             states: {
                idle: {
                    on: {
                         [EventId.UPDATE_ITEM_QUANTITY]: {
                               target: 'loading',
                               actions: [
                                    'optimisticUIUpdate',
                                    'cancelDebouncedUpdate',
                                    'updateItemBackendCall'
                              ]
                        }
                    }
               },
        },
      .....
    },
},
{
    services: {
        itemsService: (context: CheckoutContextSchema<ShippingMethodsType>, event: CheckoutEventSchema) => (cb: (event: object) => void, onReceive: any) => {
            onReceive((event: CheckoutEventSchema) => {
                if (event.type === EventId.UPDATE_ITEM_QUANTITY) {
                       // backend call, yada yada yada...
               }
       })
     },
   
    actions: {
         updateItemBackendCall: send(
            (context: CheckoutContextSchema<ShippingMethodsType>, e: CheckoutEventSchema) => ({
                type: EventId.UPDATE_ITEM_QUANTITY,
                payload: e.payload
            }),
            { to: "itemsService", delay: 1000, id: 'debounced-update' }
        ),
        cancelDebouncedUpdate: cancel('debounced-update'),
    }
}
```

Basically, I just want to debounce backend call, and in case user clicks "update quantity" button few times only request with the last value will be send. What is actually happening is that  `updateItemBackendCall` is called after 1s with the value passed in the first "click". Also - when I remove `cancelDebouncedUpdate` from actions list nothing changes. 
What am I missing here?

**Comment by davidkpiano** [comment]
✏️ Please create a [CodeSandbox (XState template)](https://codesandbox.io/s/xstate-example-template-qcwed) reproduction of this. Thanks!

**Reply by matigda** [reply]
https://codesandbox.io/s/xstate-example-template-vanilla-forked-moh67

it's here. Basically - keep clicking on button. It will show alert after 1s, counting from the first click. And with the number "2" at first. And, as you may assume, it's not what I want to achieve at the end :)

**Reply by Andarist** [reply]
This happens because you schedule a `send` action and transition immediately to the other state - so your machine gets actually to a state which **doesn't** have a transition for `UPDATE_ITEM_QUANTITY` event.

I'm not sure what exact relation do u want to have between your `items.loading` & `items.idle` states and that callback service so I can't really say how you should fix it. I hope the explanation of the problem will help you to understand the issue better and find a way to implement what you want. If it doesn't help you much - let me know, but please cover more details about the situation you'd like to achieve. I think you might be looking for a combination of delayed events and delayed transitions.

**Reply by matigda** [reply]
> This happens because you schedule a `send` action and transition immediately to the other state - so your machine gets actually to a state which **doesn't** have a transition for `UPDATE_ITEM_QUANTITY` event.
> 

THIS.
Thank you very much! That helped me with understanding and solving my issue ( at least on sandbox, I will try that in my code later :D )
